core-bundle/src/Doctrine/Backup/RetentionPolicy.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Doctrine\Backup;
+
+class RetentionPolicy implements RetentionPolicyInterface
+{
+    private int $keepMax;
+
+    /**
+     * @var array<string, \DateInterval>
+     */
+    private array $keepIntervals;
+
+    public function __construct(int $keepMax, array $keepIntervals = [])
+    {
+        $this->keepMax = $keepMax;
+        $this->keepIntervals = $this->createKeepIntervals($keepIntervals);
+    }
+
+    public function getKeepMax(): int
+    {
+        return $this->keepMax;
+    }
+
+    public function getKeepIntervals(): array
+    {
+        return $this->keepIntervals;
+    }
+
+    public function apply(Backup $latestBackup, array $allBackups): array
+    {
+        $toKeep = $allBackups;
+        $keepMax = $this->getKeepMax();
+        $keepIntervals = $this->getKeepIntervals();
+
+        // Cleanup according to days retention policy first
+        if (0 !== \count($keepIntervals)) {
+            $latestDateTime = $latestBackup->getCreatedAt();
+            $assignedPerInterval = array_fill_keys(array_keys($keepIntervals), null);
+
+            foreach ($allBackups as $k => $backup) {
+                // Do not assign the latest
+                if (0 === $k) {
+                    continue;
+                }
+
+                foreach (array_keys($assignedPerInterval) as $intervalReadable) {
+                    $interval = $this->keepIntervals[$intervalReadable];
+
+                    if (-1 === $this->compareDateIntervals($latestDateTime->diff($backup->getCreatedAt(), true), $interval)) {
+                        $assignedPerInterval[$intervalReadable] = $backup;
+                    }
+                }
+            }
+
+            // Always keep the latest
+            $toKeep = array_merge([$latestBackup], array_filter($assignedPerInterval));
+
+            // Ensure sorting again
+            usort($toKeep, static fn (Backup $a, Backup $b) => $b->getCreatedAt() <=> $a->getCreatedAt());
+        }
+
+        // Then cleanup according to maximum amount of backups to keep
+        if ($keepMax > 0) {
+            $toKeep = \array_slice($toKeep, 0, $keepMax);
+        }
+
+        return $toKeep;
+    }
+
+    /**
+     * @throws \Exception
```

Comment:
This type hint needs to be removed (see #3816).

Reply:
I've just pushed https://github.com/contao/contao/pull/3850/commits/63928ec6b1615d6bef024d23bbecdf6915035965 to already validate things in the configuration. I left the comment because it's now `public static` and you are expected to handle it. Still wrong?

Reply:
Yes, sorry. General exceptions such as `\Exception`, `\RuntimeException` or `\LogicException` are by definition always supposed to bubble up. If you want to add an exception that other developers are supposed to handle in a try/catch block, you should add (or use) a specific exception.